### DISCUSSION TITLE
Jellyfin unable to transcode using hardware acceleration, unprivileged.

### DISCUSSION THREAD

**Discussion by sheamaus** [tteck/Proxmox#2654] [Help Needed]
Hello, I've tried a few times to get this working but I am not having much success. I have had transcoding working already from within a docker container in an LXC but wanted to move Jellyfin to its own LXC. I have installed Jellyfin using the script as unpriviliged and edited the LXC.conf file the same as I have with the docker container.

Lines added to LXC.conf:

> lxc.cgroup2.devices.allow: c 226:0 rwm
lxc.cgroup2.devices.allow: c 226:128 rwm
lxc.mount.entry: /dev/dri/renderD128 dev/dri/renderD128 none bind,optional,create=file
lxc.idmap: u 0 100000 65536
lxc.idmap: g 0 100000 44
lxc.idmap: g 44 44 1
lxc.idmap: g 45 100045 62
lxc.idmap: g 107 104 1
lxc.idmap: g 108 100108 65428

Reading some of the other posts I could see that people have been having issues with renderD128 not being in the render group and did also find mine was under 'kvm' rather than render. [I followed this post](https://github.com/tteck/Proxmox/issues/2523#issuecomment-1955723524) but changed the command to `sed -i -e 's/^kvm:x:107:$/render:x:107:root,jellyfin/' -e 's/^render:x:108:root,jellyfin$/ssl-cert:x:108:/' /etc/group` and this seems to have fixed it so that the group render persists as owners in between reboots.

This hasn't fixed my issue though and I'm still having issues with videos not hardware transcoding and this is what I am finding on the logs:

> [AVHWDeviceContext @ 0x5d8b802c5f80] No VA display found for any default device.
Device creation failed: -22.
Failed to set value 'vaapi=va:,kernel_driver=i915,driver=iHD' for option 'init_hw_device': Invalid argument
Error parsing global options: Invalid argument

Also when trying to do vainfo I get:

> error: XDG_RUNTIME_DIR not set in the environment.
error: can't connect to X server!
libva info: VA-API version 1.14.0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_14
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.14 (libva 2.12.0)
vainfo: Driver version: Intel iHD driver for Intel(R) Gen Graphics - 22.3.1 ()
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            : VAEntrypointVLD
      VAProfileMPEG2Main              : VAEntrypointVLD
      VAProfileH264Main               : VAEntrypointVLD
      VAProfileH264Main               : VAEntrypointEncSliceLP
      VAProfileH264High               : VAEntrypointVLD
      VAProfileH264High               : VAEntrypointEncSliceLP
      VAProfileJPEGBaseline           : VAEntrypointVLD
      VAProfileJPEGBaseline           : VAEntrypointEncPicture
      VAProfileH264ConstrainedBaseline: VAEntrypointVLD
      VAProfileH264ConstrainedBaseline: VAEntrypointEncSliceLP
      VAProfileVP8Version0_3          : VAEntrypointVLD
      VAProfileHEVCMain               : VAEntrypointVLD
      VAProfileHEVCMain10             : VAEntrypointVLD
      VAProfileVP9Profile0            : VAEntrypointVLD
      VAProfileVP9Profile2            : VAEntrypointVLD

The vainfo output does look healthier than the response i was getting before fixing the render group issue, but there's still something not right here.

**Comment by tteck** [comment]
Try privileged to see if it works before moving to unprivileged. FYI, the `sed` command is included in the script.

**Reply by sheamaus** [reply]
Thanks, running in privileged does seem to work for hardware acceleration. 
How do I then move it to unprivileged? I tried adding `Unprvileged: 1` to the LXC config but then I can no longer access the web page or log in to the shell.

**Reply by tteck** [reply]
What's your issue with leaving the container privileged?

**Reply by sheamaus** [reply]
For the same reasons anyone would want to run a container unprivileged. To mitigate the risks from bad actors. I'd rather not have a service that goes out to the web be open to the rest of my proxmox/network. 

I've used the lines I was missing that I found in the config file created by the privileged container, and added them to a fresh install of an unprivileged container, and I'm comparing between the two looking at device ownerships. The unprivileged container before was showing kvm as the group for renderD128 but now I am getting nobody:nogroup. The same applies to card0 and fb0. Trying to chown gives me Operation not permitted errors.

**Reply by tteck** [reply]
I operate several containers with privileged access, although they are not accessible from the internet.
So, are you intending to make Jellyfin accessible from the internet?

**Reply by sheamaus** [reply]
Yes. I intend to be able to access my media away from home and allow family access.

**Reply by sheamaus** [reply]
I have figured it out now and have the steps required to complete should others come looking here for how to do this.

My only issue so far is that one of my files that had no problem transcoding inside a docker container, is transcoding video and audio, but playing without audio on any device. If you have any ideas about that it would be greatly appreciated.

Also, I'm not sure what the requirement is for the tteck script to pass through /dev/fb0, as with my testing it isn't making any difference whether it is included or not? Can any clarity be added for this?


Here are the steps I am taking to get this working in unprivileged mode:

1.  Install the script through Proxmox selecting advanced and choosing the unprivileged mode.

2. Shutdown the container and edit the containers config `nano /etc/pve/lxc/<container number>.conf` and add the following:
>lxc.cgroup2.devices.allow: a
lxc.cap.drop:
lxc.cgroup2.devices.allow: c 226:0 rwm
lxc.cgroup2.devices.allow: c 226:128 rwm
lxc.mount.entry: /dev/dri/renderD128 dev/dri/renderD128 none bind,optional,create=file
lxc.idmap: u 0 100000 65536
lxc.idmap: g 0 100000 44
lxc.idmap: g 44 44 1
lxc.idmap: g 45 100045 62
lxc.idmap: g 107 104 1
lxc.idmap: g 108 100108 65428

3. Start the container and enter this command in the shell to change GID 107 to "render" and owner for renderD128:
`sed -i -e 's/^kvm:x:107:$/render:x:107:root,jellyfin/' /etc/group`

4. Trying to install anything with 'apt install' is giving me a dpkg error, so to fix this enter the following 4 commands in the shell.
>rm /var/lib/dpkg/statoverride
rm /var/lib/dpkg/lock
dpkg --configure -a
apt-get -f install 

5. Install Intel's iGPU drivers `apt install -y intel-gpu-tools` and turn on transcoding in Jellyfin if you haven't already and that's it!

**Reply by tteck** [reply]
So, you didn't have to install `va-driver-all`, `ocl-icd-libopencl1` or `intel-opencl-icd`?
https://github.com/tteck/Proxmox/blob/6037d381ef3f061357b6d689f5fa89df26310c59/install/jellyfin-install.sh#L24-L31

**Reply by tteck** [reply]
> Install Intel's iGPU drivers `apt install -y intel-gpu-tools` 

`intel-gpu-tools` does not install any drivers. It is a collection of tools for debugging and monitoring the Intel graphics driver.

**Reply by tteck** [reply]
>lxc.idmap: g 107 104 1

This is telling the LXC to map group ID `107` inside the container to group ID `104` on the host system

>Start the container and enter this command in the shell to change GID 107 to "render"

Why is the group name `render` not set with ID `104`? Is the group `kvm` not needed?

**Reply by sheamaus** [reply]
Responding in order of the above replies.

1. Nope I didn’t have to do any of those installs and I have replicated this on fresh install of the jellyfin container. I will try adding those though to see if that will possibly help with the audio transcoding issue I’m having with the one file though and will report back later.

2. Interesting that it’s not installing any drivers, I believed it was from another post I’ve seen on the proxmox forum and this is what enabled it to work for me.

3. I’m not entirely sure on why render is changed to 107 when 44 stays the same. It has something to do with ensuring no overlap when setting up the GID pass throughs. GID kvm is not required inside the container from what I have experienced this far as you would either use it by the shell or the web UI. My source of this information was from a [Youtube video by Jim’s Garage.](https://youtu.be/0ZDr5h52OOE?si=pBdKJWpM4swgPCBY) He explains it from about 9:50.

Another step I have missed about is to add root to the render and video groups on Proxmox. I’ve edited my above instructions to include this.

**Comment by tteck** [comment]
Here's a newly posted solution.
https://forum.proxmox.com/threads/guide-jellyfin-remote-network-shares-hw-transcoding-with-intels-qsv-unprivileged-lxc.142639/

**Reply by sheamaus** [reply]
This was the post that lead me to getting the transcoding working by using `apt install -y intel-gpu-tools `

**Comment by tteck** [comment]
I'm not sure I agree with his approach. Why is he installing Jellyfin in a Docker container? Plus, take a look at the user/group settings for his renderD129 (`nobody` `_ssh`) – it seems questionable. https://youtu.be/0ZDr5h52OOE?si=TDVKkcA3ZEMGd3pY&t=960

**Reply by sheamaus** [reply]
I did see that his was showing _ssh rather than the kvm which is what mine came up with. His was built on Debian 12 though where as the tteck script is Ubuntu. I think his reason for docker would be to show something that’s easily repeatable to be deployed by the viewer and also a likely use case.

**Comment by tteck** [comment]
In my opinion, having privileges behind a well-configured proxy/tunnel seems secure enough (no tin foil hat here).
https://jellyfin.org/docs/general/administration/hardware-acceleration/intel/#lxc-on-proxmox

**Reply by sheamaus** [reply]
I think the top response in this Reddit post by a cyber security engineer makes a good point about why you probably shouldn’t run in privileged if it can be easily avoided. 
https://www.reddit.com/r/Proxmox/s/VxQHiQfgJL
